File: gui/widgets/sales_chart_widget.py
"""
Módulo de gráfico de ventas en C++ integrado con PyQt5
"""
from PyQt5.QtWidgets import QWidget
from PyQt5.QtCore import Qt, pyqtSignal, QPointF
from PyQt5.QtGui import QColor, QPainter, QPen, QFont, QBrush
import math
from typing import List, Optional

class SalesChartWidgetPy(QWidget):
    """
    Widget de gráfico de líneas para mostrar ventas.
    Implementación en Python/PyQt5 (alternativa al C++)
    """
    
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setMinimumHeight(300)
        self.setStyleSheet("background-color: white;")
        self.setAttribute(Qt.WA_StyledBackground, True)
        
        self.m_values = []
        self.m_labels = []
        self.m_title = "Gráfico de Ventas"
        self.m_lineColor = QColor(25, 118, 210)  # #1976d2
        
        # Márgenes
        self.m_marginLeft = 60
        self.m_marginRight = 40
        self.m_marginTop = 60
        self.m_marginBottom = 80
    
    def setData(self, values: List[float], labels: List[str]):
        """Establece los datos del gráfico"""
        self.m_values = values
        self.m_labels = labels
        self.update()
    
    def setTitle(self, title: str):
        """Establece el título del gráfico"""
        self.m_title = title
        self.update()
    
    def setLineColor(self, color: QColor):
        """Establece el color de la línea"""
        self.m_lineColor = color
        self.update()
    
    def clearData(self):
        """Limpia los datos del gráfico"""
        self.m_values = []
        self.m_labels = []
        self.update()
    
    def paintEvent(self, event):
        """Dibuja el gráfico"""
        try:
            painter = QPainter(self)
            painter.setRenderHint(QPainter.Antialiasing, True)
            painter.setRenderHint(QPainter.SmoothPixmapTransform, True)
            
            # Fondo blanco
            painter.fillRect(self.rect(), Qt.white)
            
            self._drawChart(painter)
        except Exception:
            # Fallback silencioso si hay error
            pass
    
    def _drawChart(self, painter: QPainter):
        """Dibuja el gráfico completo"""
        if not self.m_values:
            # Sin datos
            painter.setPen(Qt.gray)
            painter.setFont(QFont("Arial", 12))
            painter.drawText(self.rect(), Qt.AlignCenter, "No hay datos disponibles")
            return
        
        self._drawGrid(painter)
        self._drawAxes(painter)
        self._drawLine(painter)
        self._drawLabels(painter)
        self._drawLegend(painter)
    
    def _drawGrid(self, painter: QPainter):
        """Dibuja la cuadrícula"""
        chartWidth = self.width() - self.m_marginLeft - self.m_marginRight
        chartHeight = self.height() - self.m_marginTop - self.m_marginBottom
        
        painter.setPen(QPen(QColor(200, 200, 200), 1, Qt.DashLine))
        
        # Líneas horizontales
        gridLines = 5
        for i in range(gridLines + 1):
            y = self.m_marginTop + (chartHeight * i // gridLines)
            painter.drawLine(self.m_marginLeft, y, self.width() - self.m_marginRight, y)
    
    def _drawAxes(self, painter: QPainter):
        """Dibuja los ejes"""
        chartWidth = self.width() - self.m_marginLeft - self.m_marginRight
        chartHeight = self.height() - self.m_marginTop - self.m_marginBottom
        
        # Ejes principales
        painter.setPen(QPen(Qt.black, 2))
        painter.drawLine(self.m_marginLeft, self.m_marginTop, 
                        self.m_marginLeft, self.height() - self.m_marginBottom)
        painter.drawLine(self.m_marginLeft, self.height() - self.m_marginBottom,
                        self.width() - self.m_marginRight, self.height() - self.m_marginBottom)
        
        # Etiquetas de eje Y
        painter.setPen(Qt.black)
        painter.setFont(QFont("Arial", 9))
        
        if self.m_values:
            maxValue = max(self.m_values)
            if maxValue == 0:
                maxValue = 1
            
            for i in range(6):
                y = self.height() - self.m_marginBottom - (chartHeight * i // 5)
                value = int((maxValue * i) / 5)
                text_rect = self.rect()
                text_rect.setCoords(5, y - 10, self.m_marginLeft - 10, y + 10)
                painter.drawText(text_rect, Qt.AlignRight | Qt.AlignVCenter, str(value))
    
    def _drawLine(self, painter: QPainter):
        """Dibuja la línea del gráfico"""
        if not self.m_values:
            return
        
        chartWidth = self.width() - self.m_marginLeft - self.m_marginRight
        chartHeight = self.height() - self.m_marginTop - self.m_marginBottom
        
        maxValue = max(self.m_values)
        if maxValue == 0:
            maxValue = 1
        
        # Dibujar línea
        painter.setPen(QPen(self.m_lineColor, 2.4))
        
        n = len(self.m_values)
        for i in range(n - 1):
            x1 = self.m_marginLeft + (chartWidth * i) / (n - 1) if n > 1 else self.m_marginLeft
            y1 = self.height() - self.m_marginBottom - (chartHeight * self.m_values[i]) / maxValue
            
            x2 = self.m_marginLeft + (chartWidth * (i + 1)) / (n - 1) if n > 1 else self.m_marginLeft + chartWidth
            y2 = self.height() - self.m_marginBottom - (chartHeight * self.m_values[i + 1]) / maxValue
            
            painter.drawLine(int(x1), int(y1), int(x2), int(y2))
        
        # Dibujar puntos
        painter.setBrush(self.m_lineColor)
        painter.setPen(Qt.NoPen)
        
        for i in range(n):
            x = self.m_marginLeft + (chartWidth * i) / (n - 1) if n > 1 else self.m_marginLeft
            y = self.height() - self.m_marginBottom - (chartHeight * self.m_values[i]) / maxValue
            painter.drawEllipse(QPointF(x, y), 4, 4)
    
    def _drawLabels(self, painter: QPainter):
        """Dibuja las etiquetas"""
        if not self.m_values:
            return
        
        chartWidth = self.width() - self.m_marginLeft - self.m_marginRight
        
        painter.setPen(Qt.black)
        painter.setFont(QFont("Arial", 9))
        
        # Etiquetas del eje X
        n = len(self.m_labels)
        for i in range(n):
            x = self.m_marginLeft + (chartWidth * i) / (n - 1) if n > 1 else self.m_marginLeft
            y = self.height() - self.m_marginBottom + 20
            
            painter.save()
            painter.translate(x, y)
            painter.rotate(-45)
            text_rect = self.rect()
            text_rect.setCoords(0, 0, 100, 30)
            painter.drawText(text_rect, Qt.AlignLeft, self.m_labels[i])
            painter.restore()
        
        # Título
        painter.setPen(Qt.black)
        painter.setFont(QFont("Arial", 14, QFont.Bold))
        title_rect = self.rect()
        title_rect.setCoords(self.m_marginLeft, 10, self.width() - self.m_marginRight, 50)
        painter.drawText(title_rect, Qt.AlignCenter, self.m_title)
    
    def _drawLegend(self, painter: QPainter):
        """Dibuja la leyenda"""
        legX = self.width() - self.m_marginRight - 150
        legY = self.m_marginTop + 20
        legWidth = 130
        legHeight = 50
        
        # Fondo de leyenda
        painter.setPen(QPen(Qt.gray, 1))
        painter.setBrush(QColor(255, 255, 255, 240))
        painter.drawRect(legX, legY, legWidth, legHeight)
        
        # Línea de ejemplo
        painter.setPen(QPen(self.m_lineColor, 2.4))
        painter.drawLine(legX + 10, legY + 15, legX + 30, legY + 15)
        painter.setBrush(self.m_lineColor)
        painter.setPen(Qt.NoPen)
        painter.drawEllipse(QPointF(legX + 20, legY + 15), 3, 3)
        
        # Texto de leyenda
        painter.setPen(Qt.black)
        painter.setFont(QFont("Arial", 10))
        text_rect = self.rect()
        text_rect.setCoords(legX + 40, legY + 8, legX + legWidth, legY + 33)
        painter.drawText(text_rect, Qt.AlignLeft | Qt.AlignVCenter, "Ventas")

File: cpp/SalesChartWidget.cpp
#include "SalesChartWidget.h"
#include <QPainter>
#include <QResizeEvent>
#include <QPaintEvent>
#include <cmath>
#include <algorithm>

SalesChartWidget::SalesChartWidget(QWidget *parent)
    : QWidget(parent)
    , m_title("Sales Chart")
    , m_lineColor(25, 118, 210)  // Color azul (#1976d2)
    , m_marginLeft(60)
    , m_marginRight(40)
    , m_marginTop(60)
    , m_marginBottom(80)
    , m_needsRedraw(true)
{
    setMinimumHeight(300);
    setStyleSheet("background-color: white;");
    setAttribute(Qt::WA_StyledBackground, true);
}

void SalesChartWidget::setData(const std::vector<double>& values, const std::vector<QString>& labels)
{
    m_values = values;
    m_labels = labels;
    m_needsRedraw = true;
    update();
}

void SalesChartWidget::setTitle(const QString& title)
{
    m_title = title;
    m_needsRedraw = true;
    update();
}

void SalesChartWidget::setLineColor(const QColor& color)
{
    m_lineColor = color;
    m_needsRedraw = true;
    update();
}

void SalesChartWidget::clearData()
{
    m_values.clear();
    m_labels.clear();
    m_needsRedraw = true;
    update();
}

void SalesChartWidget::paintEvent(QPaintEvent *event)
{
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing, true);
    painter.setRenderHint(QPainter::SmoothPixmapTransform, true);
    
    // Fondo blanco
    painter.fillRect(rect(), Qt::white);
    
    drawChart(painter);
}

void SalesChartWidget::resizeEvent(QResizeEvent *event)
{
    m_needsRedraw = true;
    QWidget::resizeEvent(event);
}

void SalesChartWidget::drawChart(QPainter &painter)
{
    if (m_values.empty()) {
        // Dibujar mensaje si no hay datos
        painter.setPen(Qt::gray);
        painter.setFont(QFont("Arial", 12));
        painter.drawText(rect(), Qt::AlignCenter, "No hay datos disponibles");
        return;
    }
    
    drawGrid(painter);
    drawAxes(painter);
    drawLine(painter);
    drawLabels(painter);
    drawLegend(painter);
}

void SalesChartWidget::drawGrid(QPainter &painter)
{
    // Área útil del gráfico
    int chartWidth = width() - m_marginLeft - m_marginRight;
    int chartHeight = height() - m_marginTop - m_marginBottom;
    
    // Dibujar líneas de cuadrícula
    painter.setPen(QPen(QColor(200, 200, 200), 1, Qt::DashLine));
    
    // Líneas horizontales
    int gridLines = 5;
    for (int i = 0; i <= gridLines; ++i) {
        int y = m_marginTop + (chartHeight * i / gridLines);
        painter.drawLine(m_marginLeft, y, width() - m_marginRight, y);
    }
}

void SalesChartWidget::drawAxes(QPainter &painter)
{
    int chartWidth = width() - m_marginLeft - m_marginRight;
    int chartHeight = height() - m_marginTop - m_marginBottom;
    
    // Ejes
    painter.setPen(QPen(Qt::black, 2));
    painter.drawLine(m_marginLeft, m_marginTop, m_marginLeft, height() - m_marginBottom);
    painter.drawLine(m_marginLeft, height() - m_marginBottom, width() - m_marginRight, height() - m_marginBottom);
    
    // Etiquetas de eje Y
    painter.setPen(Qt::black);
    painter.setFont(QFont("Arial", 9));
    
    if (!m_values.empty()) {
        double maxValue = *std::max_element(m_values.begin(), m_values.end());
        if (maxValue == 0) maxValue = 1;
        
        for (int i = 0; i <= 5; ++i) {
            int y = height() - m_marginBottom - (chartHeight * i / 5);
            double value = (maxValue * i) / 5;
            painter.drawText(QRect(5, y - 10, m_marginLeft - 10, 20), 
                           Qt::AlignRight | Qt::AlignVCenter, 
                           QString::number(static_cast<int>(value)));
        }
    }
}

void SalesChartWidget::drawLine(QPainter &painter)
{
    if (m_values.empty()) return;
    
    int chartWidth = width() - m_marginLeft - m_marginRight;
    int chartHeight = height() - m_marginTop - m_marginBottom;
    
    double maxValue = *std::max_element(m_values.begin(), m_values.end());
    if (maxValue == 0) maxValue = 1;
    
    painter.setPen(QPen(m_lineColor, 2.4));
    
    // Dibujar línea
    for (size_t i = 0; i < m_values.size() - 1; ++i) {
        double x1 = m_marginLeft + (chartWidth * i) / (m_values.size() - 1);
        double y1 = height() - m_marginBottom - (chartHeight * m_values[i]) / maxValue;
        
        double x2 = m_marginLeft + (chartWidth * (i + 1)) / (m_values.size() - 1);
        double y2 = height() - m_marginBottom - (chartHeight * m_values[i + 1]) / maxValue;
        
        painter.drawLine(QPointF(x1, y1), QPointF(x2, y2));
    }
    
    // Dibujar puntos
    painter.setBrush(m_lineColor);
    painter.setPen(Qt::NoPen);
    for (size_t i = 0; i < m_values.size(); ++i) {
        double x = m_marginLeft + (chartWidth * i) / (m_values.size() - 1);
        double y = height() - m_marginBottom - (chartHeight * m_values[i]) / maxValue;
        painter.drawEllipse(QPointF(x, y), 4, 4);
    }
}

void SalesChartWidget::drawLabels(QPainter &painter)
{
    if (m_values.empty()) return;
    
    int chartWidth = width() - m_marginLeft - m_marginRight;
    
    painter.setPen(Qt::black);
    painter.setFont(QFont("Arial", 9));
    
    // Etiquetas del eje X
    for (size_t i = 0; i < m_labels.size(); ++i) {
        double x = m_marginLeft + (chartWidth * i) / (m_values.size() - 1);
        int y = height() - m_marginBottom + 20;
        
        painter.save();
        painter.translate(x, y);
        painter.rotate(-45);
        painter.drawText(QRect(0, 0, 100, 30), Qt::AlignLeft, m_labels[i]);
        painter.restore();
    }
    
    // Título
    painter.setPen(Qt::black);
    painter.setFont(QFont("Arial", 14, QFont::Bold));
    painter.drawText(QRect(m_marginLeft, 10, width() - m_marginLeft - m_marginRight, 40), 
                    Qt::AlignCenter, m_title);
}

void SalesChartWidget::drawLegend(QPainter &painter)
{
    // Rectángulo de leyenda
    int legX = width() - m_marginRight - 150;
    int legY = m_marginTop + 20;
    int legWidth = 130;
    int legHeight = 50;
    
    painter.setPen(QPen(Qt::gray, 1));
    painter.setBrush(QColor(255, 255, 255, 240));
    painter.drawRect(legX, legY, legWidth, legHeight);
    
    // Línea de ejemplo
    painter.setPen(QPen(m_lineColor, 2.4));
    painter.drawLine(legX + 10, legY + 15, legX + 30, legY + 15);
    painter.setBrush(m_lineColor);
    painter.setPen(Qt::NoPen);
    painter.drawEllipse(QPointF(legX + 20, legY + 15), 3, 3);
    
    // Texto de leyenda
    painter.setPen(Qt::black);
    painter.setFont(QFont("Arial", 10));
    painter.drawText(QRect(legX + 40, legY + 8, legWidth - 50, 25), 
                    Qt::AlignLeft | Qt::AlignVCenter, "Ventas");
}
